docs/fpc-key-dist.puml

```diff
@@ -1,97 +1,245 @@
-**ECC**
-+ easy to securely access m protected through TLCC
-- difficult with application policy
-TBD ...
-end note
+  Admin1        -> ERCC1            ++: query registered enclave's key
+    note right
+        ERCC will likely return the list of enclaves with their
+        cryptographic material. Admin1 then selects the non-provisioned one(s).
+    end note
+  return ECC2_S_PK /'Admin1 -> ERCC1'/
+
+  Admin1        -> ECC_Manager1     ++: DistributeKey(CC_name, CC_ver, CH_id, ECC2_S_PK)
+  ECC_Manager1  -> ECC1             ++: DistributeKey(CC_name, CC_ver, CH_id, ECC2_S_PK)
+  ECC1          -> Enclave1         ++: DistributeKey(CC_name, CC_ver, CH_id, ECC2_S_PK)
+  Enclave1      -> ERCC1            ++: GetRegistration(ECC2_S_PK)
+  return registration_message <- {CC_name, CC_ver, CH_id, org_id, ECC2_S_PK, ECC2_E_PK, attestation_report} /'Enclave1 -> ERCC1'/
+  Enclave1      -> TLCC_Enclave1    ++: check_committed_data(registration_message)
+  return true /'Enclave1 -> TLCC_Enclave1'/
+    note left
+        if false, abort
+    end note
+  Enclave1      -> TLCC_Enclave1    ++: can_endorse(ECC2_S_PK)
+    note right
+        TLCC checks the endorsement policy to determine
+        if the recipient enclave is an eligible endorser.
+    end note
+    note right
+        TLCC should check that the recipient enclave's MRENCLAVE
+        matches the chaincode definition's MRENCLAVE.
+        This could enable key distribution from an old cc version,
+        with a different MRENCLAVE.
+    end note
+  return true /'Enclave1 -> TLCC_Enclave1'/
+    note left
+        if false, abort
+    end note
+  Enclave1      -> Enclave1         :   check enclave chaincode definition against sealed data
+  Enclave1      -> Enclave1         :   check ECC2_S_PK, ECC_E_PK, attestation_report
+  Enclave1      -> Enclave1         :   m <- encrypt {SEK, CC_SK} with ECC2_E_PK
+    note left
+        enclave encrypts state encryption key and chaincode private key
+        - m denotes encrypted message
+    end note
+
+  Enclave1      -> Enclave1         :   SIG_ECC1 <- sign {ECC1_S_PK, ECC2_s_PK, m, CC_PK} with ECC1_S_SK
+    note left
+        enclave signs the key material it distributes together with the
+        sender and intended recipient identity
+    end note
+
+  return ECC1_S_PK, ECC2_s_PK, m, CC_PK, SIG_ECC1 /'ECC1 -> Enclave1'/
+  return ECC1_S_PK, ECC2_s_PK, m, CC_PK, SIG_ECC1 /'ECC_Manager1 -> ECC1'/
+  return ECC1_S_PK, ECC2_s_PK, m, CC_PK, SIG_ECC1 /'Admin1 -> ECC_Manager1'/
+
+  note right Admin1
+    Key distribution is in progress. We need a means to transfer the message with
+    the encrypted keys to the recipient enclave. The next step uses the ledger
+    itself for this data transfer. This additionally allows to log the two phases
+    of the key distribution: (1) the sender enclave produced the encrypted data blob;
+    (2) the recipient enclave correctly received it. This can be useful when dealing
+    with enclaves of two different orgs.
```

Comment:
Do we allow re-keying?

Reply:
re-keying (in the sense of replacing the state encryption key) not yet.